cmd/deploy/input_test.go

```diff
@@ -2,9 +2,276 @@ package main
+	getDigest := func(d ociregistry.Digest) *ociregistry.Digest {
+		return &d
+	}
+	for _, x := range []struct {
+		name       string
+		rawLookup  map[string]string
+		wantLookup map[ociregistry.Digest]registry.Reference
+		wantDigest *ociregistry.Digest
+		wantErr    string
+	}{
+		{
+			name: "single digest",
+			rawLookup: map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "example:foo",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": {
+					Host:       "docker.io",
+					Repository: "library/example",
+					Tag:        "foo",
+					Digest:     "sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+				},
+			},
+			wantDigest: getDigest("sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d"),
+			wantErr:    "",
+		},
+		{
+			name: "multiple digests",
+			rawLookup: map[string]string{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": "example:foo",
+				"sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86": "foo/bar:baz",
+				"sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e": "localhost:5000/example:bar",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d": {
+					Host:       "docker.io",
+					Repository: "library/example",
+					Tag:        "foo",
+					Digest:     "sha256:9ef42f1d602fb423fad935aac1caa0cfdbce1ad7edce64d080a4eb7b13f7cd9d",
+				},
+				"sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86": {
+					Host:       "docker.io",
+					Repository: "foo/bar",
+					Tag:        "baz",
+					Digest:     "sha256:0cb474919526d040392883b84e5babb65a149cc605b89b117781ab94e88a5e86",
+				},
+				"sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e": {
+					Host:       "localhost:5000",
+					Repository: "example",
+					Tag:        "bar",
+					Digest:     "sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+				},
+			},
+			wantDigest: nil,
+			wantErr:    "",
+		},
+		{
+			name: "basic fallback",
+			rawLookup: map[string]string{
+				"": "example",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"": {
+					Host:       "docker.io",
+					Repository: "library/example",
+				},
+			},
+			wantDigest: nil,
+			wantErr:    "",
+		},
+		{
+			name: "Digest fallback",
+			rawLookup: map[string]string{
+				"": "example@sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+			},
+			wantLookup: map[ociregistry.Digest]registry.Reference{
+				"": {
+					Host:       "docker.io",
+					Repository: "library/example",
+					Digest:     "sha256:25be82253336f0b8c4347bc4ecbbcdc85d0e0f118ccf8dc2e119c0a47a0a486e",
+				},
+			},
+			wantDigest: getDigest(""),
```

Comment:
Oh, this one is surprising -- the current implementation handles it correctly, so probably a `TODO` for the future that this ought to return the digest directly (and only have the empty string fallback for tag-lookup) :+1: